### PR TITLE
QRadar better retry, fix duplicates and polling

### PR DESCRIPTION
**Pull request by ilaner**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-hq.paloaltonetworks.local/browse/XSUP-15649?filter=-1
https://jira-hq.paloaltonetworks.local/browse/XSUP-13239
https://jira-hq.paloaltonetworks.local/browse/XSUP-15397?filter=-1
https://jira-hq.paloaltonetworks.local/browse/XSUP-15244?filter=-1

## Description
- Improved retry mechanism handling of failed events fetch.
- Improved missing events messages in the QRadar widgets.
- Fixed an error message when **create-search** command failed.
- Fixed duplicate incidents creation.
- Try to poll events immediately after creating a search in **qradar-search-retrieve-events** command.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 1]
```diff
@@ -3400,14 +3400,15 @@ def create_events_search(client: Client,
                          ) -> str:
     additional_where = ''' AND LOGSOURCETYPENAME(devicetype) = 'Custom Rule Engine' ''' \
         if fetch_mode == FetchMode.correlations_events_only.value else ''
+    now: int = int(datetime.now().timestamp() * 1000)
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3403]:
```suggestion
    now_millis: int = int(datetime.now().timestamp() * 1000)
```
or `now_milliseconds`

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 2]
```diff
@@ -3400,14 +3400,15 @@ def create_events_search(client: Client,
...
     try:
         # Get all the events starting from one hour after epoch
         if not offense_start_time:
             offense = client.offenses_list(offense_id=offense_id)
             offense_start_time = offense['start_time']
         query_expression = (
             f'SELECT {events_columns} FROM events WHERE INOFFENSE({offense_id}) {additional_where} limit {events_limit} '
-            f'START {offense_start_time}'
+            f'START {offense_start_time} STOP {now}'
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3411]:
```suggestion
            f'START {offense_start_time} STOP {now_millis}'
```

**Review by tkatzir — APPROVED** [review]
Looks good.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 3]
```diff
@@ -1642,21 +1640,24 @@ def enrich_offense_with_events(client: Client, offense: Dict, fetch_mode: str, e
...
+        else:
+            events, failure_message = poll_offense_events_with_retry(client, search_id, int(offense_id))
+        events_fetched = sum(int(event.get('eventcount', 1)) for event in events)
+        print_debug_msg(f'Events fetched for offense {offense_id}: {events_fetched}/{events_count}.')
+        offense['events_fetched'] = events_fetched
+        if events:
+            offense['events'] = events
+            break
+        time.sleep(FAILURE_SLEEP)
+        print_debug_msg(f'No events for offense {offense_id}. Retrying.')
```
Comment (DeanArbel) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1658]: Are there no `events` or were there none fetched?

Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1658]: none fetched, fixed message

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 ''' ADVANCED GLOBAL PARAMETERS '''
 
-FAILURE_SLEEP = 15  # sleep between consecutive failures events fetch
+FAILURE_SLEEP = 20  # sleep between consecutive failures events fetch
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:20]: Shouldn't this be 30 seconds, to round up if needed?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 5]
```diff
@@ -1642,22 +1640,25 @@ def enrich_offense_with_events(client: Client, offense: Dict, fetch_mode: str, e
...
+        offense['events_fetched'] = events_fetched
+        if events:
+            offense['events'] = events
+            break
+        print_debug_msg(f'No events were fetched for offense {offense_id}. Retrying in {FAILURE_SLEEP}.')
+        time.sleep(FAILURE_SLEEP)
     else:
-        print_debug_msg(f'No events were fetched for offense {offense_id}'
-                        f'Adding to mirroring queue to be queried again.')
+        print_debug_msg(f'No events were fetched for offense {offense_id}.\n'
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1660]:
```suggestion
        print_debug_msg(f'No events were fetched for offense {offense_id}. '
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 6]
```diff
@@ -1537,9 +1537,7 @@ def create_search_with_retry(client: Client, fetch_mode: str, offense: Dict, eve
     offense_id = offense['id']
     while num_of_failures <= max_retries:
         try:
-            ret_value = create_events_search(client, fetch_mode, event_columns, events_limit, offense_id, offense['start_time'])
-            time.sleep(EVENTS_MODIFIED_SECS)
-            return ret_value
+            return create_events_search(client, fetch_mode, event_columns, events_limit, offense_id, offense['start_time'])
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1540]: From what I could tell, `create_events_search` doesn't raise an exception,
so you'll only iterate once?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 7]
```diff
@@ -1537,9 +1537,7 @@ def create_search_with_retry(client: Client, fetch_mode: str, offense: Dict, eve
     offense_id = offense['id']
     while num_of_failures <= max_retries:
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1538]: This:
`if num_of_failures == max_retries:`
(and its block) should be out of the loop.

**Review by tkatzir — COMMENTED** [review]
I believe I see some potential issues with the loop.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py (hunk 4 above)
Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:20]: 20 seconds should be enough (the third request should succeed in all cases)

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 8]
```diff
@@ -1533,23 +1538,14 @@ def create_search_with_retry(client: Client, fetch_mode: str, offense: Dict, eve
...
-                            f'Retry number {num_of_failures}/{max_retries}.')
+                            f'Retry number {i+1}/{max_retries}.')
             print_debug_msg(traceback.format_exc())
-            num_of_failures += 1
-            if num_of_failures == max_retries:
-                print_debug_msg(f'Max retries for creating search for offense: {offense_id}. Returning empty.')
-                break
-            time.sleep(FAILURE_SLEEP)
-    print_debug_msg(f'Could not create search query for {offense_id}.')
+    print_debug_msg(f'Max retries for creating search for offense: {offense_id}. Returning empty.')
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1548]:
```suggestion
    print_debug_msg(f'Reached max retries for creating a search for offense: {offense_id}. Returning error.')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 9]
```diff
@@ -1642,22 +1638,25 @@ def enrich_offense_with_events(client: Client, offense: Dict, fetch_mode: str, e
...
+            failure_message = 'Search for events was failed.'
+        else:
+            events, failure_message = poll_offense_events_with_retry(client, search_id, int(offense_id))
+        events_fetched = sum(int(event.get('eventcount', 1)) for event in events)
+        print_debug_msg(f'Events fetched for offense {offense_id}: {events_fetched}/{events_count}.')
+        offense['events_fetched'] = events_fetched
+        if events:
+            offense['events'] = events
+            break
+        print_debug_msg(f'No events were fetched for offense {offense_id}. Retrying in {FAILURE_SLEEP}.')
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1655]:
```suggestion
        print_debug_msg(f'No events were fetched for offense {offense_id}. Retrying in {FAILURE_SLEEP} seconds.')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 10]
```diff
@@ -1642,22 +1638,25 @@ def enrich_offense_with_events(client: Client, offense: Dict, fetch_mode: str, e
...
+        time.sleep(FAILURE_SLEEP)
     else:
-        print_debug_msg(f'No events were fetched for offense {offense_id}'
-                        f'Adding to mirroring queue to be queried again.')
+        print_debug_msg(f'No events were fetched for offense {offense_id}. '
+                        f'If mirroring is enabled, it will be queried again in mirroring.')
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1659]: What if mirroring isn't enabled?
Should we advise the user to enable mirroring?

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 11]
```diff
@@ -1874,12 +1873,14 @@ def perform_long_running_loop(client: Client, offenses_per_fetch: int, fetch_mod
...
         # if incident creation fails, it'll drop the data and try again in the next iteration
-        demisto.createIncidents(incidents)
-
         safely_update_context_data(context_data=context_data,
                                    version=ctx_version,
                                    should_update_last_fetch=True)
 
+        demisto.createIncidents(incidents)
+        print_debug_msg(
+            f'Successfully Created {len(incidents)} incidents. Incidents created {[incident["name"] for incident in incidents]}')
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1882]:
```suggestion
            f'Successfully Created {len(incidents)} incidents. Incidents created: {[incident["name"] for incident in incidents]}')
```

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 12]
```diff
@@ -2408,6 +2412,9 @@ def qradar_search_create_command(client: Client, params: Dict, args: Dict) -> Co
                                         int(offense_id),
                                         start_time,
                                         return_raw_response=True)
+        if response == QueryStatus.ERROR.value:
+            raise DemistoException(f'Could not create search for offense_id: {offense_id}')
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:2416]:
```suggestion
            raise DemistoException(f'Could not create events search for offense_id: {offense_id}')
```
Distinguish the exception and log messages.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py [hunk 13]
```diff
@@ -3538,7 +3527,12 @@ def qradar_search_retrieve_events_command(client: Client,
         next_run_in_seconds=interval_in_secs,
         args=polling_args,
     )
-    return CommandResults(scheduled_command=scheduled_command)
+    return CommandResults(scheduled_command=scheduled_command,
+                          readable_output=f'Search ID: {search_id}',
+                          outputs_prefix='QRadar.SearchEvents',
+                          outputs_key_field='ID',
+                          outputs=search_command_results.outputs if search_command_results else None
```
Comment (tkatzir) [inline Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:3534]:
```suggestion
                          outputs=search_command_results.outputs if search_command_results else None,
```

**Review by tkatzir — COMMENTED** [review]
Looks good!
Waiting with approval for when it's ready.

File: Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py (hunk 10 above)
Comment (ilaner) [reply Packs/QRadar/Integrations/QRadar_v3/QRadar_v3.py:1659]: If mirroring isn't enabled, the events will not be fetched at all. We already advise enabling mirroring in this case

**Review by tkatzir — APPROVED** [review]
Looks good to me, os I've approved,
But please wait for @DeanArbel's approval as well, before merging.

**Review by DeanArbel — APPROVED** [review]
Good job, see my 2 comments

File: Packs/QRadar/ReleaseNotes/2_2_10.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+- Improved retry mechanism handling of failed events fetch.
+- Improved missing events messages in the QRadar widgets.
+- Fixed an error message when **qradar-search-create** command failed.
+- Fixed duplicate incidents creation.
+- Try to poll events immediately after creating a search in **qradar-search-retrieve-events** command.
+
+#### Scripts
+##### QRadarMirroringEventsStatus
+- Improved missing events messages in the QRadar widgets.
+- Updated the Docker image to: *demisto/python3:3.10.5.31928*.
```
Comment (DeanArbel) [inline Packs/QRadar/ReleaseNotes/2_2_10.md:13]:
```suggestion
- Improved retry handling when not all events are fetched.
- QRadar widgets now display better messages when not all events are fetched.
- Fixed an error message when **qradar-search-create** command failed.
- Fixed an issue where duplicate incidents were created.
- The command **qradar-search-retrieve-events** will try to poll events immediately after creating a search.

#### Scripts
##### QRadarMirroringEventsStatus
- Improved messages when not all events are fetched.
- Updated the Docker image to: *demisto/python3:3.10.5.31928*.
```

File: Packs/QRadar/ReleaseNotes/2_2_10.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
+
+#### Integrations
+##### IBM QRadar v3
+- Updated the Docker image to: *demisto/python3:3.10.5.33323*.
+- Improved retry handling when not all events are fetched.
+- QRadar widgets now display better messages when not all events are fetched.
+- Fixed an error message when **qradar-search-create** command failed.
```
Comment (DeanArbel) [inline Packs/QRadar/ReleaseNotes/2_2_10.md:7]:
```suggestion
- Improved the error message when **qradar-search-create** command fails.
```

**Pull request merged**